### ISSUE TITLE
Discussion: remove seek/rewind interface for hummock iterator

### ISSUE THREAD

**Issue by hzxa21** [risingwavelabs/risingwave#7441]
Hummock iterator provides [seek](https://github.com/risingwavelabs/risingwave/blob/cbb29601aeedd1c10a8d22ff8771f7ccf9d574fa/src/storage/src/hummock/iterator/mod.rs#L131) and [rewind](https://github.com/risingwavelabs/risingwave/blob/cbb29601aeedd1c10a8d22ff8771f7ccf9d574fa/src/storage/src/hummock/iterator/mod.rs#L122) interfaces. However, [a stream interface](https://github.com/risingwavelabs/risingwave/blob/cbb29601aeedd1c10a8d22ff8771f7ccf9d574fa/src/storage/src/store.rs#L39) is provided to the streaming/batch executors without exposing these interfaces, which makes us think maintaining these two interfaces is unneccessary.

Currently the usage of hummock iterator is a bit werid since we will call seek/rewind when creating the iterator before wrapping it in a iterator stream and giving back to the executor. I think we need to make a choice between the above options:
1. Keep the seek/rewind interfaces in hummock iterator and expose it to the executors. In other words, the executor will call seek/rewind explicitly after the iterator is created.
2. Remove the seek/rewind interfaces in hummock iterator and keep using the iterator stream in executors.

Since I don't see a use case that an executor needs to seek multiple times on the same iterator, I personally prefer 2) since we have seen burdens on mataining seek/rewind when doing refactor and optimizations. cc @Little-Wallace @soundOfDestiny @wenym1 @fuyufjh @BugenZhao @TennyZhuang

**Comment by st1page** [comment]
I like to remove it to follow the KISS principle. But also, I think we need a benchmark to tell us the overhead to construct a new iter compared with the iter::seek.

**Comment by BugenZhao** [comment]
> 1. the executor will call seek/rewind explicitly after the iterator is created.

This might be not elegant and hard to use in Rust. 🤔 For example,
- If we've turned the iterator into a generator stream, we can't obtain it back.
- We may need to add these methods in the upper-level iterators, like tables.
- There's lifetime in the mem-table iterator, if we want to hold and reuse it, we can't write into the mem-table. 🥵

If the iterator creation becomes a bottleneck and we don't care about the performance gain from the compaction in the same epoch, I'm considering whether it works if we simply add some memoization for the same epoch. 🤔

**Comment by st1page** [comment]
> There's lifetime in the mem-table iterator, if we want to hold and reuse it, we can't write into the mem-table. hot_face

We can not write anything into the table anyway, or the iter's result will be inconsistent. 

> If we've turned the iterator into a generator stream, we can't obtain it back. So we may need to say goodbye to the stream ecosystem

Let's consider what will the concurrent read optimization. Currently, we are just use `buffer_unordered` to drive the futures concurrently. If want to optimize it with `seek`, we might need to maintain an "iter_pool" and reset the pool when we need to write on the table or get a barrier to update the epoch.

**Comment by st1page** [comment]
> If the iterator creation becomes a bottleneck and we don't care about the performance gain from the compaction in the same epoch, I'm considering whether it works if we simply add some memoization for the same epoch.

just to cache an "IteratorBuilder" in the local state store and when write or flush happens invalid it?

**Comment by BugenZhao** [comment]
> We can not write anything into the table anyway, or the iter's result will be inconsistent.

But this is really the case we want to optimize, right? The inconsistency is brought by the memtable iterator, while the storage part stays the same and can be reused. Or you need to break the encapsulation for state table iterator.

**Comment by hzxa21** [comment]
> > 1. the executor will call seek/rewind explicitly after the iterator is created.
> 
> This might be not elegant and hard to use in Rust. 🤔 For example,
> 
> * If we've turned the iterator into a generator stream, we can't obtain it back. So we may need to say goodbye to the stream ecosystem.
> * We may need to add these methods in the upper-level iterators, like tables.
> * There's lifetime in the mem-table iterator, if we want to hold and reuse it, we can't write into the mem-table. 🥵



Agreed. It is definitly a big refactoring if we choose option 1 and like you said I am not even sure whether it is possible. This is also one reason why I prefer option 2 (given that we have already fully adopted Rust stream for storage iterator and we didn't see performance issue).

>
> If the iterator creation becomes a bottleneck and we don't care about the performance gain from the compaction in the same epoch, I'm considering whether it works if we simply add some memoization for the same epoch. 🤔

Right. in our current implementation, we mainly do three things on iterator creation:
1. SST pruning based on key range, epoch and table id
2. SST pruning based on prefix hint (requires fetching SST meta block)
3. iterator seek

1 can be very efficient and 2 can also be efficient with a good meta cache. So I guess if there is a bottleneck in iterator creation, it mostly comes from 3 and letting executor call seek/rewind won't help. 

Also, we can only reuse a iterator if we want to read data from the same epoch and provide a wide key range to create the iterator. This requirement is somehow strict and can make prefetch/SST pruning more tricky because storage won't know the exact key range on iterator creation.

**Comment by BugenZhao** [comment]
> Also, we can only reuse a iterator if we want to read data from the same epoch and provide a wide key range to create the iterator. This requirement is somehow strict and can make prefetch/SST pruning more tricky because storage won't know the exact key range on iterator creation.

Makes sense. This is a big change for both upper-level usages and lower-level optimization.

**Comment by hzxa21** [comment]
@wenym1 Given that mutable memtable can impl `rewind` and `seek`, this issue can be de-prioritized. But I think it is still good-to-have.

**Comment by soundOfDestiny** [comment]
> @wenym1 Given that mutable memtable can impl `rewind` and `seek`, this issue can be de-prioritized. But I think it is still good-to-have.

yeah we hardly need to re-use a hummock iterator
